File: 백준/Silver/2579. 계단 오르기/계단 오르기.c
#include <stdio.h>
#define MAX(a, b) (a>b?a:b)

int stair[301];
int dp[301];

int max_sum(int);

int main(void)
{
    int n;
    scanf("%d", &n);
    for(int i=1;i<=n;i++){
        scanf("%d", &stair[i]);
    }
    printf("%d\n", max_sum(n));
    return 0;
}
int max_sum(int n)
{
    //초기값 기록
    dp[0]=0;
    dp[1]=stair[1];
    dp[2]=stair[1]+stair[2];
    for(int i=3;i<=n;i++){
        //1층아래, 2층아래 합 중 최댓값 선택
        //1층아래 합은 연속된 층이 아닌 값이 되게 3층아래 합+1층아래 값
        dp[i]=MAX((dp[i-3]+stair[i-1]), dp[i-2])+stair[i];
    }
    return dp[n];
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
def max_sum(n):
    dp[1] = scores[1]
    dp[2] = dp[1] + scores[2]
    for i in range(3, n+1):
        # i층_최대점수 = max((i-1층_최대점수), (i-2층_최대점수)) + i층_점수
        # i-1층_최대점수는 i-2에서 올라오지 않아야 하므로 i-3에서 올라온 경우로 변환
        dp[i] = max(dp[i-3]+scores[i-1], dp[i-2]) + scores[i]
    return dp[n]

n = int(input())
scores =[0] + [int(input()) for _ in range(n)]
dp = [0] * (n+1)
if len(scores) < 3:
    print(sum(scores))
else:
    print(max_sum(n))